cloudstack/resource_cloudstack_instance.go

```diff
@@ -136,6 +137,23 @@ func resourceCloudStackInstance() *schema.Resource {
 				},
 			},
 
+			"user_data_base64": &schema.Schema{
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"user_data"},
+				StateFunc: func(v interface{}) string {
+					return fmt.Sprintf("%x", sha1.Sum([]byte(v.(string))))
+				},
+				ValidateFunc: func(v interface{}, name string) (warns []string, errs []error) {
+					if _, err := base64.StdEncoding.DecodeString(v.(string)); err != nil {
+						errs = append(errs, fmt.Errorf(
+							"%s: must be base64-encoded", name,
+						))
+					}
+					return
+				},
+			},
```

Comment:
I wonder if the API for this would be better/cleaner is we just update the docs for the existing `user_data` field to make it clear that you can pass either plain text or base64 encoded text to this field.

Then the rest of the logic around the `user_data` could remain the same, except for the `getUserData` method which should first detect if the content is already base64 encoded and only encode the content it if its not.

Reply:
That would work too but i looked at the aws provider and it has a dedicated field:

https://www.terraform.io/docs/providers/aws/r/instance.html#user_data_base64

Reply:
I understand. I guess both are valid options, but I really prefer to to reuse the existing field as I think it's a cleaner design (both in the code as in the UX).